ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/LSM.hs

```diff
@@ -0,0 +1,462 @@
+  , Session
+  , LSM.openSession
+  , LSM.closeSession
+  , stdMkBlockIOFS
+
+    -- * snapshot-converter
+  , implTakeSnapshot
+  ) where
+
+import Cardano.Binary as CBOR
+import Codec.Serialise (decode)
+import qualified Control.Monad as Monad
+import Control.Monad.Trans (lift)
+import Control.Monad.Trans.Except
+import Control.ResourceRegistry
+import Control.Tracer
+import qualified Data.Foldable as Foldable
+import Data.Functor.Contravariant ((>$<))
+import qualified Data.List as List
+import qualified Data.Map.Strict as Map
+import Data.Maybe
+import Data.MemPack
+import qualified Data.Primitive.ByteArray as PBA
+import qualified Data.Set as Set
+import Data.String (fromString)
+import qualified Data.Text as Text
+import qualified Data.Vector as V
+import qualified Data.Vector.Mutable as VM
+import qualified Data.Vector.Primitive as VP
+import Data.Void
+import Database.LSMTree (Session, Table)
+import qualified Database.LSMTree as LSM
+import NoThunks.Class
+import Ouroboros.Consensus.Block
+import Ouroboros.Consensus.Config
+import Ouroboros.Consensus.Ledger.Abstract
+import Ouroboros.Consensus.Ledger.Extended
+import Ouroboros.Consensus.Ledger.SupportsProtocol
+import qualified Ouroboros.Consensus.Ledger.Tables.Diff as Diff
+import Ouroboros.Consensus.Ledger.Tables.Utils
+import Ouroboros.Consensus.Storage.LedgerDB.API
+import Ouroboros.Consensus.Storage.LedgerDB.Args
+import Ouroboros.Consensus.Storage.LedgerDB.Snapshots
+import Ouroboros.Consensus.Storage.LedgerDB.TraceEvent
+import qualified Ouroboros.Consensus.Storage.LedgerDB.V2.Args as V2
+import Ouroboros.Consensus.Storage.LedgerDB.V2.LedgerSeq
+import Ouroboros.Consensus.Util (chunks)
+import Ouroboros.Consensus.Util.Args
+import Ouroboros.Consensus.Util.CRC
+import Ouroboros.Consensus.Util.Enclose
+import Ouroboros.Consensus.Util.IOLike
+import Ouroboros.Consensus.Util.IndexedMemPack
+import System.FS.API
+import qualified System.FS.BlockIO.API as BIO
+import System.FS.BlockIO.IO
+import Prelude hiding (read)
+
+-- | Type alias for convenience
+type UTxOTable m = Table m TxInBytes TxOutBytes Void
+
+instance NoThunks (Table m txin txout Void) where
+  showTypeOf _ = "Table"
+  wNoThunks _ _ = pure Nothing
+
+data LSMClosedExn = LSMClosedExn
+  deriving (Show, Exception)
+
+{-------------------------------------------------------------------------------
+  TxOuts
+-------------------------------------------------------------------------------}
+
+newtype TxOutBytes = TxOutBytes {unTxOutBytes :: LSM.RawBytes}
+
+toTxOutBytes :: IndexedMemPack (l EmptyMK) (TxOut l) => l EmptyMK -> TxOut l -> TxOutBytes
+toTxOutBytes st txout =
+  let barr = indexedPackByteArray True st txout
+   in TxOutBytes $ LSM.RawBytes (VP.Vector 0 (PBA.sizeofByteArray barr) barr)
+
+fromTxOutBytes :: IndexedMemPack (l EmptyMK) (TxOut l) => l EmptyMK -> TxOutBytes -> TxOut l
+fromTxOutBytes st (TxOutBytes (LSM.RawBytes (VP.force -> (VP.Vector _ _ barr)))) =
+  case indexedUnpackLeftOver' st barr of
```

Comment:
Is it really expected that we `VP.force` *and* allow leftovers?

More generally (as also briefly discussed on Slack some time ago), it seems rather easy to just invoke `mempack` in such a way that it is aware of the offset and length of the `ByteArray`, meaning that we can avoid the copy via `VP.force`.

Same idea, different guise: We could write an appropriate [`Buffer`](https://hackage-content.haskell.org/package/mempack-0.1.2.0/docs/Data-MemPack-Buffer.html#t:Buffer) instance for `RawBytes`.